library/kani_core/src/mem_init.rs

```diff
@@ -343,5 +362,54 @@ macro_rules! kani_mem_init {
         fn copy_init_state_single<const LAYOUT_SIZE: usize, T>(from: *const T, to: *const T) {
             copy_init_state::<LAYOUT_SIZE, T>(from, to, 1);
         }
+
+        /// Information about currently tracked argument, used for passing union initialization
+        /// state across function boundaries. This struct is written to by the caller and read from
+        /// by the callee.
+        #[derive(Clone, Copy)]
+        struct ArgumentBuffer {
+            selected_argument: usize,
+            saved_address: *const (),
+            layout_size: usize,
+        }
+
+        /// Non-deterministically store information about currently tracked argument in the argument
+        /// buffer.
+        #[kanitool::disable_checks(pointer)]
+        #[rustc_diagnostic_item = "KaniStoreArgument"]
+        fn store_argument<const LAYOUT_SIZE: usize, T>(from: *const T, selected_argument: usize) {
+            let (from_ptr, _) = from.to_raw_parts();
+            let should_store: bool = super::any();
+            if should_store {
+                unsafe {
+                    ARGUMENT_BUFFER = Some(ArgumentBuffer {
+                        selected_argument,
+                        saved_address: from_ptr,
+                        layout_size: LAYOUT_SIZE,
+                    })
+                }
+            }
+        }
+
+        /// Load information from the argument buffer (if the argument position matches) via copying
+        /// the memory initialization information from an address in the caller to an address in the
+        /// callee. Otherwise, mark that the argument as initialized, as it will be checked by
+        /// another non-deterministic branch. Reset the argument buffer after loading from it.
+        #[kanitool::disable_checks(pointer)]
+        #[rustc_diagnostic_item = "KaniLoadArgument"]
+        fn load_argument<const LAYOUT_SIZE: usize, T>(to: *const T, selected_argument: usize) {
+            let (to_ptr, _) = to.to_raw_parts();
+            unsafe {
```

Comment:
since this function is disabling pointer checks, it would be very nice if we could minimize the usage of `unsafe`.